docs/design/audit-v2.md

```diff
@@ -0,0 +1,74 @@
+# Auditing V2: Random Node Selection
+
+## Abstract
+
+This design doc outlines how we will implement a Version 2 of the auditing service.
+With our current auditing service, we are currently auditing per segment of data.
+We need to keep this existing method of auditing, but also add an additional audit selection process that selects random nodes to audit.
+
+## Background
+
+As our network grows, it will take longer for nodes to get vetted.
+This is because every time an upload happens, we send 5% of the uploaded data to unvetted nodes and 95% to vetted nodes.
+When auditing occurs, we currently select a random stripe within a segment.
+If we're selecting audits at random per byte, every segment has some percentage that went to vetted nodes.
+As more nodes join the network, it will become exponentially less likely that an unvetted node will be audited since most data will be stored on vetted nodes.
+With a satellite with one petabyte of data currently, new nodes will take one month to get vetted, which is way too long.
+
+We want a way to select segments to audit based on statistically randomly selected storage nodes.
+However, the way we select nodes should also be biased toward unvetted nodes.
+Currently, there's not a way to select a stripe based on a storage node.
+
+## Design
+
+Two different loops will select audits:
+- One loop will select based on stripes.
+- The second loop will select based on nodes.
+
+After selection, the rest of the auditing process will occur the same way (picking a segment, picking a stripe, downloading all erasure shares associated with that stripe).
+The chances of selecting the same stripe are rare, but should be accounted for.
+
+With both loops, we should have auditing that occurs statistically uniform across both nodes and bytes.
+
+### **Selection via Reservoir Sampling**
+
+Reservoir sampling: a family of algorithms for randomly choosing a sample of items with uniform probability from a stream of data (of unknown size).
+
+During uploads or during garbage collection bloom filter generation, we're going through all segments and creating filters per node.
```

Comment:
Could we do reservoir sampling during the segment-based audit loop? [And thus actually only have one loop?]

Reply:
oh yeah I guess the gc observer is just going through the pointerdb same as the existing audit loop. I feel like there's some difference I'm missing too but I'm not sure.. 🤔

Reply:
The segment based audit loop uses random selection so it will not uniformly hit all nodes/segments. The GC observer (or any observer on the metainfo loop) will always hit every segment on the satellite exactly once, allowing us to get the most accurate possible sample for each node. In garbage collection, we also keep track of the current number of pieces seen for each node ID, which would allow us to easily determine the probability that we should replace something in that node's reservoir when we find a new piece for that node

Reply:
As discussed in standup, we'll consider using a single audit loop.  That means we'll have to update the main audit to use the metainfo observer loop, and use a reservoir sample to get random segments for the main loop.